### DISCUSSION TITLE
Customizable asset position in <head>

### DISCUSSION THREAD

**Discussion by AngusMorton** [withastro/roadmap#444] [Proposal]
As a developer, I want to improve the performance of my pages by optimising the page load and removing waterfalls for Server Side Rendering. Currently, because Astro appends assets at the end of the `<head>` tag we end up with unnecessary page load waterfalls.

### Problem

The way Astro assets are injected into the `<head>` tag causes page load waterfalls for slow-rendering pages. For example, this is how we currently render the `<head>`:

```astro
<head>
    <!-- Slot calls an API and adds SEO meta tags. --->
    <slot name="async-head-that-blocks" />

    <!-- Astro assets -->
    <link rel="stylesheet" href="/_astro/page.e1119834.css">
</head>

// Page load waterfall:
// # - - - - - HTML - - - - -  |
//                              #  - - - Styles - - - |
//                                                     | Page Load
```

While we're waiting for the `<slot>` to render we could be streaming the stylesheets and the users browser could be loading them.

So, ideally, it would be rendered with the blocking `<slot>` tag at the end:

```astro
<head>
    <!-- Astro assets. -->
    <link rel="stylesheet" href="/_astro/page.e1119834.css">

    <!-- Slot calls an API and adds SEO meta tags. -->
    <slot name="async-head-that-blocks" />
</head>

// Page load waterfall:
// # - - - - - HTML - - - - -  |
//    #  - - - Styles - - - |                                        
//                              | Page Load
```

### Proposal

One possible solution would be to use a marker in the head for where you want Astro to insert its assets. For example, we could introduce a new `<astro-assets />` tag that when rendering, if we encounter it, we replace it with the assets. If we don't find an `<astro-assets />` tag then Astro inserts its assets at the end of the head by default.

```
// Layout.astro
<head>
    <!-- Astro assets. -->
    <astro-assets />

    <!-- Slot calls an API and adds SEO meta tags. -->
    <slot name="async-head-that-blocks" />
</head>
```

**Comment by matthewp** [comment]
This would be one solution: https://github.com/withastro/roadmap/discussions/499
